src/lib/src/grid3d/grid.cpp

```diff
@@ -421,4 +421,146 @@ get_bounding_box(const Grid &grid)
     return std::make_tuple(minv, maxv);
 }
 
+inline double
+interp1d(const double x1, const double x2, const double fx)
+{
+    return x1 + fx * (x2 - x1);
+};
+
+inline xyz::Point
+interp(const xyz::Point pt1, const xyz::Point pt2, const double f)
+{
+    return xyz::Point{ pt1.x + f * (pt2.x - pt1.x), pt1.y + f * (pt2.y - pt1.y),
+                       pt1.z + f * (pt2.z - pt1.z) };
+}
+
```

Comment:
It is difficult to follow the flow here. I would propose that you utilise existing `refine_vertically` for the vertical refinement, and then make a separate function `refine_horizontally` which possibly takes one axis at the time (i.e. being generic). You can eithermake a new instance of a Grid in cpp, or steer that from python.